tutorials/appengine-pusher/index.md

```diff
@@ -0,0 +1,449 @@
+    for (String pair : pairs) {
+      int idx = pair.indexOf("=");
+      query_pairs.put(
+          URLDecoder.decode(pair.substring(0, idx), "UTF-8"),
+          URLDecoder.decode(pair.substring(idx + 1), "UTF-8"));
+    }
+    return query_pairs;
+  }
+}
+```
+
+### Send messages over a channel
+
+Authenticated clients can submit messages to the channel over HTTP using a server-side endpoint
+as shown in the following example.
+
+The sender can be [excluded](https://pusher.com/docs/server_api_guide/server_excluding_recipients)
+from receiving the broadcast message , by passing in its own socket ID when triggering an event.
+Individual messages are limited to 10KB in size.
+
+For more on publishing messages to multiple channels or batching multiple messages, refer to
+[this resource](https://pusher.com/docs/server_api_guide/interact_rest_api#publishing-events).
+
+[embedmd]:# (java/src/main/java/com/example/appengine/pusher/SendMessageServlet.java /public class/ $)
+```java
+public class SendMessageServlet extends HttpServlet {
+
+  private Gson gson = new GsonBuilder().create();
+  private TypeReference<Map<String, String>> typeReference =
+      new TypeReference<Map<String, String>>() {};
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Parse POST request body received in the format :
+    // [{"message": "my-message", "socket_id": "1232.24", "channel": "presence-my-channel"}]
+
+    String body = CharStreams.readLines(request.getReader()).toString();
+    String json = body.replaceFirst("^\\[", "").replaceFirst("\\]$", "");
+    Map<String, String> data = gson.fromJson(json, typeReference.getType());
+    String message = data.get("message");
+    String socketId = data.get("socket_id");
+    String channelId = data.get("channel_id");
+
+    User user = UserServiceFactory.getUserService().getCurrentUser();
+    // user email prefix as display name for current logged in user
+    String displayName = user.getNickname().replaceFirst("@.*", "");
+
+    // Create a message including the user email prefix to display in the chat window
+    String taggedMessage = "<strong>&lt;" + displayName + "&gt;</strong> " + message;
+    Map<String, String> messageData = new HashMap<>();
+    messageData.put("message", taggedMessage);
+
+    // Send a message over the Pusher channel (maximum size of a message is 10KB)
+    Result result =
+        PusherService.getDefaultInstance()
+            .trigger(
+                channelId,
+                "new_message", // name of event
+                messageData,
+                socketId); // (optional) use client socket_id to exclude the sender from receiving the message
+
+    // result.getStatus() == SUCCESS indicates successful transmission
+    messageData.put("status", result.getStatus().name());
+
+    response.getWriter().println(gson.toJson(messageData));
+  }
+}
+```
+
+## Using Pusher on your client
+
```

Comment:
Avoid nesting headings. Add a sentence here explaining what the reader will learn. Something better than this: "The following sections explain how to use Pusher on your client..."